### PR TITLE
event: Reduce potential for lock contention while executing dispatcher post callbacks.

### PR DESCRIPTION
**Pull request by antoniovicente**
Risk Level: low, minor code refactor
Testing: n/a, minimal functional changes expected.
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: source/common/event/dispatcher_impl.cc
Comment (jmarantz) [inline source/common/event/dispatcher_impl.cc:270]: I didn't think std::move defined clearly what state `post_callbacks_` would be in after this statement. So I was thinking maybe swap would be better. WDYT?

File: source/common/event/dispatcher_impl.cc
Comment (jmarantz) [inline source/common/event/dispatcher_impl.cc:274]: is there some reason not to write this as:
```
for (auto& callback : callbacks) {
  callback();
}
```

**Review by jmarantz — COMMENTED** [review]
this looks great; just a couple of nits. Did we wind up finding any evidence of contention?

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc:274]: Differences in deletion order for the callback objects.

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc:270]: post_callbacks_ is a list, so I think this is well defined.  I could add post_callbacks_.clear(); after the move.

File: source/common/event/dispatcher_impl.cc
Comment (jmarantz) [reply source/common/event/dispatcher_impl.cc:274]: good point.  add that to the comment?

File: source/common/event/dispatcher_impl.cc
Comment (jmarantz) [reply source/common/event/dispatcher_impl.cc:270]: ok that seems like it would de-risk it. From https://en.cppreference.com/w/cpp/utility/move

> Unless otherwise specified, all standard library objects that have been moved from are placed in a valid but unspecified state. That is, only the functions without preconditions, such as the assignment operator, can be safely used on the object after it was moved from: ... `str.clear(); // OK, clear() has no preconditions`

is that better than swap? I'm fine either way.

I did not see anything about std::list's move semantics in https://en.cppreference.com/w/cpp/container/list

Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc:270]: std::swap is defined as 3 move assignments.  Added the clear and a comment.

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc:277]: Added a comment and test.

I also wanted to test yield behavior, but it turns out apparently ordering of schedulable callback scheduling is not deterministic so the yield behavior is not testable.

**Review by jmarantz — APPROVED** [review]
@envoyproxy/senior-maintainers ptal

File: source/common/event/dispatcher_impl.cc
Comment (ggreenway) [reply source/common/event/dispatcher_impl.cc:270]: Instead of the `clear()`, I think it would be better to `ASSERT(post_callbacks_.empty())`. Then if something gets messed up and it somehow became a copy instead of a move (due to c++-shenanigans or something) we'll notice quickly, instead of callbacks not getting run.

File: source/common/event/dispatcher_impl.cc
Comment (ggreenway) [inline source/common/event/dispatcher_impl.cc:267]: I think this case can be deleted; the function behaves correctly without it AFAICT.

**Review by ggreenway — CHANGES_REQUESTED** [review]
I like the change; much better to not lock/unlock the mutex for each callback separately.

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc:267]: Done.

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc]: Done.

I was thinking of it as a micro optimization but didn't realize that most post_callbacks_ is usually non-empty when we get here.

File: source/common/event/dispatcher_impl.cc
Comment (chradcliffe) [reply source/common/event/dispatcher_impl.cc:267]: Just jumping in here... what about:
```c++
std::list<std::function<void()>> callbacks = [this]() {
  Thread::LockGuard lock(post_lock_);
  return std::exchange(post_callbacks_, {});
}();
```
or some variation using `std::exchange`. The pattern of using exchange is mentioned in reference to an event Dispatcher class in [an article](https://www.fluentcpp.com/2020/09/25/stdexchange-patterns-fast-safe-expressive-and-probably-underused/) on fluentcpp.com.

Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc:267]: I went through the article.  I think the most relevant section is the "Why not just move?".  The issue of expressing the "empty after move" constraint is addressed here by means of the ASSERT(post_callbacks_,empty());

File: source/common/event/dispatcher_impl.cc
Comment (lizan) [inline source/common/event/dispatcher_impl.cc:274]: can be in oneline: `callbacks.front().callback()`?

**Review by lizan — COMMENTED** [review]
This changed the behavior that if a posted callback post another callback, it will be in next event cycle instead of the current one. Is that intended?

File: source/common/event/dispatcher_impl.cc
Comment (ggreenway) [inline source/common/event/dispatcher_impl.cc:276]: nit: capitalize `Pop`.

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc:274]: I think the one line version would be:
```
callbacks.front()();
```
or
```
callbacks.front().operator();
```

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc]: Done.

**Review by antoniovicente — COMMENTED** [review]
> This changed the behavior that if a posted callback post another callback, it will be in next event cycle instead of the current one. Is that intended?

Yes, there is a change in post loop behavior.  The loop yields after the group of callbacks present is executed.  Post callbacks added by the first set of post callbacks execute in the same event loop iteration (thanks to post_cb_->scheduleCallbackCurrentIteration()) but after other loop events have a chance to execute.

This change in behavior is desirable.  I wanted to add a test to cover this behavior but I ran into an issue with ordering of post callbacks vs schedulable callbacks.  I thought order was deterministic but apparently it isn't.  The yield behavior will be testable as a consequence of https://github.com/envoyproxy/envoy/pull/14293

File: test/common/event/dispatcher_impl_test.cc
Comment (antoniovicente) [inline test/common/event/dispatcher_impl_test.cc:404]: @lizan Added test coverage for yield behavior.

**Review by ggreenway — APPROVED** [review]
LGTM, but please wait until @lizan responds before merging.

File: source/common/event/dispatcher_impl.cc
Comment (lizan) [reply source/common/event/dispatcher_impl.cc:279]: yeah, either way, just style nit.

**Review by lizan — APPROVED** [review]
Just wanted to double-check the behavior change.

No strong preference on the nit, up to you.

File: source/common/event/dispatcher_impl.cc
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.cc]: Done.

**Review by lizan — APPROVED** [review]

**Pull request merged**
